File: server/api/src/me/eddiep/ghost/game/match/world/physics/Hitbox.java
package me.eddiep.ghost.game.match.world.physics;

import me.eddiep.ghost.utils.Vector2f;
import me.eddiep.ghost.utils.VectorUtils;

public class Hitbox {
    private Polygon bounds;
    private double rotation;
    private String name;

    public Hitbox(String name, Polygon bounds) {
        this.bounds = bounds;
        this.name = name;
    }

    public Hitbox(String name, Vector2f... points) {
        this.bounds = new Polygon(points);
        this.name = name;
    }

    public Polygon getPolygon() {
        return bounds;
    }

    public String getName() {
        return name;
    }

    public boolean isPointInside(Vector2f point) {
        return VectorUtils.isPointInside(point, bounds.getPoints());
    }

    public boolean isHitboxInside(Hitbox hitbox) {
        for (Face face : hitbox.getPolygon().getFaces()) {
            for (Face face2 : bounds.getFaces()) {
                boolean isIntersecting = VectorUtils.lineIntersects(face.getPointA(), face.getPointB(), face2.getPointA(), face2.getPointB());
                if (isIntersecting)
                    return true;
            }
        }

        return false;
    }
}

File: client/GhostClient/Core/Sharp2D_API/Hitbox.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Microsoft.Xna.Framework;

namespace Ghost.Core.Sharp2D_API
{
    public class Hitbox
    {
        private Polygon _polygon;
        private string name;

        public string Name
        {
            get { return name; }
        }

        public Polygon Polygon
        {
            get { return _polygon; }
        }

        public Hitbox(string name, Polygon bounds)
        {
            _polygon = bounds;
            this.name = name;
        }

        public Hitbox(string name, params Vector2[] points)
        {
            _polygon = new Polygon(points);
            this.name = name;
        }

        public bool isPointInside(Vector2 point)
        {
            return Vector2Utils.isPointInside(point, _polygon.Points);
        }

        public bool isHitboxInside(Hitbox hitbox)
        {
            foreach (Vector2 point in hitbox.Polygon.Points)
            {
                if (Vector2Utils.isPointInside(point, Polygon.Points))
                {
                    return true;
                }
            }
            foreach (Face face in hitbox.Polygon.Faces)
            {
                foreach (Face face2 in _polygon.Faces)
                {
                    bool isIntersecting = Vector2Utils.lineIntersects(face.PointA, face.PointB, face2.PointA,
                        face2.PointB);
                    if (isIntersecting)
                        return true;
                }
            }

            return false;
        }
    }
}
